### DISCUSSION TITLE
Support with-statement contexts for mock engines

### DISCUSSION THREAD

**Discussion by jakebrinkmann** [sqlalchemy/sqlalchemy#7461] [Usage Questions]
### Describe the use case

I am using `pytest` and I want to do some testing that avoids actual transactions in the database. I would like to use `create_mock_engine` and monkeypatch the real engine. 

However, I get an exception: `AttributeError: __enter__`, I think because the mock connection does not support this context? https://github.com/sqlalchemy/sqlalchemy/blob/8d28bd8d4a0acc209884d29729020506a46f8bd4/lib/sqlalchemy/engine/mock.py#L14

### Databases / Backends / Drivers targeted

mock

### Example Use

```python
import pytest
from sqlalchemy import create_mock_engine

@pytest.fixture
def mock_engine(mocker):
    def executor(sql, *multiparams, **params):
        print(sql.compile(dialect=engine.dialect))

    engine = create_mock_engine("postgresql://", executor)
    mocker.patch("myproj.orm.engine", engine)
```

In my test subject, I am using the context:

```python
def list_records() -> list:
    stmt = select(OrmModel)
    with engine.connect() as conn:
        records = conn.execute(stmt).mappings().all()
    return records
```

**Comment by jakebrinkmann** [comment]
For now, I'll just roll my own `MockConnection`:

```python
class MockConnection(object):
    def connect(self, **kwargs):
        return self

    def execute(self, object_, *multiparams, **params):
        print(str(object_))
        return self

    def __enter__(self):
        return self

    def __exit__(self, type_, value, traceback):
        return self
```

**Comment by zzzeek** [comment]
the mock engine is mis-named; it isn't intended for "mocks" as used in test suites.  It's a facade over a dialect that is used to generate DDL statements as strings.  It does not have much relevance for a test suite unless you are testing that your DDL generated for a certain model is what's expected.

To make mocks that simulate database transactions is a different thing altogether.    The hard part is that queries have to return results, so it requires that you create a system by which actual results can themselves be mocked, meaning you have to mock out Result and everything else you expect to get and populate these mocks with the data you'd expect to receive for each test.  This is a complex task that is better suited to code you'd roll yourself in your application and/or a third party library.   The idea is that if you are mocking out database code such that it isn't talking to a database at all, you no longer are going to be sending SQL strings to a database, because there is none.   So you don't need a dialect, you just need a system by which sending query A will get you results B.

I did do a proof of concept of this system years ago, which was oriented towards the now-legacy Query object, you can see it here: https://bitbucket.org/zzzeek/pycon2014_atmcraft/src/master/  (sorry it's not moved to github yet), the mock result system is here: https://bitbucket.org/zzzeek/pycon2014_atmcraft/src/master/atmcraft/tests/_mock_session.py  and examples of populating it with data and using it is here: https://bitbucket.org/zzzeek/pycon2014_atmcraft/src/master/atmcraft/tests/test_views_mockdb.py .  

the above code is kind of mind-bending but it works, but in practice I think it's much more pragmatic to just use a SQLite database for test suites, this is what people normally do.
